In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Load data

### Subtask:
Muat data dari file CSV ke dalam DataFrame pandas.


In [2]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Test_dataset/Artikel FTMM - Sheet1.csv')
display(df.head())

title  \
0  Desa Binaan 2.0 Wujud Nyata Pengabdian Mahasis...   
1  Forum Group Discussion SEGTA 2025 Bahas Commun...   
2  Solar Harvest: Pengering Panen Cerdas Berbasis...   
3  BEM FTMM x ARGON Kembangkan Pupuk Limbah Sisik...   
4  Pemasangan SmartAqua oleh RASENA UNAIR Dorong ...   

                                         description  
0  FTMM NEWS – Badan Eksekutif Mahasiswa Fakultas...  
1  FTMM NEWS – Sesudah kegiatan orientasi pada Se...  
2  FTMM NEWS – Tim Inovasi Ikatan Mahasiswa Elekt...  
3  FTMM NEWS –  BEM FTMM Universitas Airlangga me...  
4  FTMM NEWS –Tim RASENA Universitas Airlangga ke...

## Text cleaning

### Subtask:
Lakukan pembersihan teks menggunakan Regular Expression untuk menghapus HTML tag, angka, simbol, atau tautan.


In [3]:
import re

def clean_text(text):
    # Remove HTML tags
    text = re.sub(r'<.*?>', '', text)
    # Remove URLs or links
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)
    # Remove numbers
    text = re.sub(r'\d+', '', text)
    # Remove symbols or special characters (keeping spaces)
    text = re.sub(r'[^\w\s]', '', text)
    # Convert to lowercase
    text = text.lower()
    # Remove leading/trailing whitespace
    text = text.strip()
    return text

# Potong 11 karakter awal dulu
df['description_trimmed'] = df['description'].str[11:]

# Baru lakukan cleaning
df['cleaned_title'] = df['title'].apply(clean_text)
df['cleaned_description'] = df['description_trimmed'].apply(clean_text)

display(df.head())


title  \
0  Desa Binaan 2.0 Wujud Nyata Pengabdian Mahasis...   
1  Forum Group Discussion SEGTA 2025 Bahas Commun...   
2  Solar Harvest: Pengering Panen Cerdas Berbasis...   
3  BEM FTMM x ARGON Kembangkan Pupuk Limbah Sisik...   
4  Pemasangan SmartAqua oleh RASENA UNAIR Dorong ...   

                                         description  \
0  FTMM NEWS – Badan Eksekutif Mahasiswa Fakultas...   
1  FTMM NEWS – Sesudah kegiatan orientasi pada Se...   
2  FTMM NEWS – Tim Inovasi Ikatan Mahasiswa Elekt...   
3  FTMM NEWS –  BEM FTMM Universitas Airlangga me...   
4  FTMM NEWS –Tim RASENA Universitas Airlangga ke...   

                                 description_trimmed  \
0   Badan Eksekutif Mahasiswa Fakultas Teknologi ...   
1   Sesudah kegiatan orientasi pada Senin (14/07)...   
2   Tim Inovasi Ikatan Mahasiswa Elektro (IME) Un...   
3    BEM FTMM Universitas Airlangga melalui Kemen...   
4  Tim RASENA Universitas Airlangga kembali melak...   

                                       cleaned_title  \
0  desa binaan  wujud nyata pengabdian mahasiswa ...   
1  forum group discussion segta  bahas community ...   
2  solar harvest pengering panen cerdas berbasis ...   
3  bem ftmm x argon kembangkan pupuk limbah sisik...   
4  pemasangan smartaqua oleh rasena unair dorong ...   

                                 cleaned_description  
0  badan eksekutif mahasiswa fakultas teknologi m...  
1  sesudah kegiatan orientasi pada senin  lalu pe...  
2  tim inovasi ikatan mahasiswa elektro ime unive...  
3  bem ftmm universitas airlangga melalui kemente...  
4  tim rasena universitas airlangga kembali melak...

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   title                      30 non-null     object
 1   description                30 non-null     object
 2   description_trimmed        30 non-null     object
 3   cleaned_title              30 non-null     object
 4   cleaned_description        30 non-null     object
 5   processed_description      30 non-null     object
 6   processed_description_str  30 non-null     object
dtypes: object(7)
memory usage: 1.8+ KB


## Text preprocessing

### Subtask:
Lakukan tokenisasi kata, stopword removal, dan stemming/lemmatization.


In [4]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

# Download necessary NLTK data, including punkt_tab
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True)
nltk.download('stopwords', quiet=True)

# Using a standard list of Indonesian stopwords
stop_words = set(stopwords.words('indonesian'))

def tokenize_text(text):
    # Handle potential None values
    if text is None:
        return []
    return word_tokenize(text)

def remove_stopwords(tokens):
    return [word for word in tokens if word not in stop_words]

# Using Porter Stemmer for stemming
stemmer = PorterStemmer()
def stem_tokens(tokens):
    return [stemmer.stem(word) for word in tokens]

df['processed_description'] = df['cleaned_description'].apply(tokenize_text).apply(remove_stopwords).apply(stem_tokens)

display(df.head())

title  \
0  Desa Binaan 2.0 Wujud Nyata Pengabdian Mahasis...   
1  Forum Group Discussion SEGTA 2025 Bahas Commun...   
2  Solar Harvest: Pengering Panen Cerdas Berbasis...   
3  BEM FTMM x ARGON Kembangkan Pupuk Limbah Sisik...   
4  Pemasangan SmartAqua oleh RASENA UNAIR Dorong ...   

                                         description  \
0  FTMM NEWS – Badan Eksekutif Mahasiswa Fakultas...   
1  FTMM NEWS – Sesudah kegiatan orientasi pada Se...   
2  FTMM NEWS – Tim Inovasi Ikatan Mahasiswa Elekt...   
3  FTMM NEWS –  BEM FTMM Universitas Airlangga me...   
4  FTMM NEWS –Tim RASENA Universitas Airlangga ke...   

                                 description_trimmed  \
0   Badan Eksekutif Mahasiswa Fakultas Teknologi ...   
1   Sesudah kegiatan orientasi pada Senin (14/07)...   
2   Tim Inovasi Ikatan Mahasiswa Elektro (IME) Un...   
3    BEM FTMM Universitas Airlangga melalui Kemen...   
4  Tim RASENA Universitas Airlangga kembali melak...   

                                       cleaned_title  \
0  desa binaan  wujud nyata pengabdian mahasiswa ...   
1  forum group discussion segta  bahas community ...   
2  solar harvest pengering panen cerdas berbasis ...   
3  bem ftmm x argon kembangkan pupuk limbah sisik...   
4  pemasangan smartaqua oleh rasena unair dorong ...   

                                 cleaned_description  \
0  badan eksekutif mahasiswa fakultas teknologi m...   
1  sesudah kegiatan orientasi pada senin  lalu pe...   
2  tim inovasi ikatan mahasiswa elektro ime unive...   
3  bem ftmm universitas airlangga melalui kemente...   
4  tim rasena universitas airlangga kembali melak...   

                               processed_description  
0  [badan, eksekutif, mahasiswa, fakulta, teknolo...  
1  [kegiatan, orientasi, senin, peserta, sustain,...  
2  [tim, inovasi, ikatan, mahasiswa, elektro, ime...  
3  [bem, ftmm, universita, airlangga, kementerian...  
4  [tim, rasena, universita, airlangga, melaksana...

## Vector representation

### Subtask:
Buat representasi vektor dokumen menggunakan TF, IDF, dan TF-IDF.


In [5]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Join the list of tokens back into strings for TfidfVectorizer
df['processed_description_str'] = df['processed_description'].apply(lambda tokens: ' '.join(tokens) if isinstance(tokens, list) else '')

# Instantiate TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()

# Fit and transform the processed text data
tfidf_matrix = tfidf_vectorizer.fit_transform(df['processed_description_str'])

# Print the shape of the TF-IDF matrix
print("Shape of TF-IDF matrix:", tfidf_matrix.shape)

# Get the list of feature names (words)
feature_names = tfidf_vectorizer.get_feature_names_out()
print("Number of features (words):", len(feature_names))
# Optionally print some feature names
# print("Feature names (first 20):", feature_names[:20])

Shape of TF-IDF matrix: (30, 1766)
Number of features (words): 1766


## Visualize important words

### Subtask:
Visualisasikan 10 kata paling penting (TF-IDF tertinggi) dari setiap artikel.


In [6]:
import pandas as pd

# Get the feature names (words) from the tfidf_vectorizer
feature_names = tfidf_vectorizer.get_feature_names_out()

# Create a DataFrame from the tfidf_matrix
# The matrix is sparse, convert to dense array for DataFrame creation
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=feature_names, index=df.index)

# For each row (article) in the TF-IDF DataFrame:
print("Top 10 TF-IDF words for each article:")
for index, row in tfidf_df.iterrows():
    # Sort the TF-IDF values in descending order and select top 10
    top_words = row.sort_values(ascending=False).head(10)

    # Print the article index and the top 10 words with their scores
    print(f"\nArticle {index}:")
    for word, score in top_words.items():
        print(f"  {word}: {score:.4f}")

Top 10 TF-IDF words for each article:

Article 0:
  volunt: 0.3094
  binaan: 0.3024
  desa: 0.2808
  program: 0.1998
  minggu: 0.1856
  bem: 0.1748
  dusun: 0.1646
  karangploso: 0.1512
  masyarakat: 0.1361
  filtrasi: 0.1237

Article 1:
  topik: 0.3145
  gili: 0.2352
  segta: 0.2097
  to: 0.1764
  iyang: 0.1764
  and: 0.1675
  green: 0.1573
  in: 0.1573
  qualiti: 0.1406
  sustain: 0.1406

Article 2:
  harvest: 0.3102
  ime: 0.2766
  panen: 0.2440
  solar: 0.2342
  alat: 0.2318
  kepuh: 0.1551
  modern: 0.1551
  wangkal: 0.1551
  hasil: 0.1550
  tim: 0.1488

Article 3:
  pupuk: 0.4213
  nanoshield: 0.2487
  agro: 0.2487
  petani: 0.2319
  sisik: 0.1865
  kimia: 0.1865
  kesuburan: 0.1865
  tanah: 0.1855
  poin: 0.1518
  dusun: 0.1392

Article 4:
  tambak: 0.3720
  smartaqua: 0.3188
  alat: 0.2851
  rasena: 0.2428
  gunung: 0.2428
  anyar: 0.2250
  pakan: 0.2126
  pengguna: 0.1788
  memantau: 0.1457
  tim: 0.1430

Article 5:
  sqym: 0.4164
  tanah: 0.2472
  lahan: 0.2456
  tani: 0.2323

## Pos tagging and ner

### Subtask:
Lakukan POS Tagging dan NER pada salah satu artikel.


In [7]:
!pip install stanza

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 14.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 590.6/590.6 kB 18.4 MB/s eta 0:00:00


In [8]:
import stanza
stanza.download('id')

INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Downloading default packages for language: id (Indonesian) ...


INFO:stanza:Downloaded file to /root/stanza_resources/id/default.zip
INFO:stanza:Finished downloading models and saved to /root/stanza_resources


In [9]:
nlp = stanza.Pipeline(lang='id', processors='tokenize, pos, lemma')

INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: id (Indonesian):
| Processor | Package      |
----------------------------
| tokenize  | gsd          |
| mwt       | gsd          |
| pos       | gsd_charlm   |
| lemma     | gsd_nocharlm |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: pos
INFO:stanza:Loading: lemma
INFO:stanza:Done loading processors!


In [10]:
# Select the text from the first row of the 'cleaned_description' column
article_text = df['cleaned_description'].iloc[0]

# Process the text with the nlp pipeline
doc = nlp(article_text)

# Print the results for POS tagging and lemma
print("POS Tagging and Lemma for the first article:")
for sentence in doc.sentences:
    print("\nSentence:")
    for word in sentence.words:
        # Stanza's word object contains lemma and pos information
        print(f"Word: {word.text}\tLemma: {word.lemma}\tPOS: {word.pos}")

POS Tagging and Lemma for the first article:

Sentence:
Word: badan	Lemma: badan	POS: NOUN
Word: eksekutif	Lemma: eksekutif	POS: ADJ
Word: mahasiswa	Lemma: mahasiswa	POS: NOUN
Word: fakultas	Lemma: fakultas	POS: NOUN
Word: teknologi	Lemma: teknologi	POS: NOUN
Word: maju	Lemma: maju	POS: VERB
Word: dan	Lemma: dan	POS: CCONJ
Word: multidisiplin	Lemma: multidisiplin	POS: NOUN
Word: universitas	Lemma: universitas	POS: PROPN
Word: airlangga	Lemma: airlangga	POS: NOUN
Word: melaksanakan	Lemma: laksana	POS: VERB
Word: program	Lemma: program	POS: NOUN
Word: desa	Lemma: desa	POS: NOUN
Word: binaan	Lemma: bina	POS: NOUN
Word: sebagai	Lemma: sebagai	POS: ADP
Word: wujud	Lemma: wujud	POS: NOUN
Word: nyata	Lemma: nyata	POS: NOUN
Word: pengabdian	Lemma: abdi	POS: NOUN
Word: masyarakat	Lemma: masyarakat	POS: NOUN
Word: program	Lemma: program	POS: NOUN
Word: tahunan	Lemma: tahun	POS: NOUN
Word: bem	Lemma: bem	POS: NOUN
Word: ftmm	Lemma: ftmm	POS: NOUN
Word: ini	Lemma: ini	POS: DET
Word: berfokus	Lemma

In [ ]:
!pip install -q transformers datasets spacy torch

<h1>NER</H1>

In [36]:
# Install libraries

# Download IndoBERT NER model dan tokenizer
from transformers import AutoTokenizer, AutoModelForTokenClassification
from transformers import pipeline

# Model IndoBERT untuk NER (fine-tuned)
model_checkpoint = "indobenchmark/indobert-base-p1"

tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)
model = AutoModelForTokenClassification.from_pretrained("cahya/bert-base-indonesian-NER")

# Buat pipeline NER
nlp_ner = pipeline("ner", model=model, tokenizer=tokenizer, aggregation_strategy="simple")

# Jalankan NER
entities = nlp_ner(article_text)

# Tampilkan hasil
for ent in entities:
    print(f"{ent['word']} ({ent['entity_group']}) : {ent['score']:.2f}")


Some weights of the model checkpoint at cahya/bert-base-indonesian-NER were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu
Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


mahasiswa fakultas teknologi maju (PRD) : 0.39
##iplin universitas airlangga (PRD) : 0.51
program desa binaan (PRD) : 0.27
masyarakat program tahunan bem ft (ORG) : 0.36
ini (PRD) : 0.42
pada (ORG) : 0.26
pemberdayaan masyarakat desa dan tahun ini berkolaborasi dengan (PRD) : 0.45
argo (CRD) : 0.32
##n (PRD) : 0.21
kegiatan (ORG) : 0.32
ini (PRD) : 0.53
untuk (PER) : 0.43
melalui (ORG) : 0.30
pendekatan (PRD) : 0.35
teknologi (ORG) : 0.45
dan peningkatan kemandirian desa dengan (PRD) : 0.47
strategi tersebut desa (PRD) : 0.30
binaan (ORG) : 0.26
bagi (ORG) : 0.12
pengembangan (PRD) : 0.30
program (ORG) : 0.24
di (CRD) : 0.50
wilayah lain (QTY) : 0.25
pertama kegiatan (ORG) : 0.26
ini berlangsung pada hari jumat hingga (PRD) : 0.40
di (CRD) : 0.58
desa (QTY) : 0.32
##ong kabupaten pasuruan rombongan (PRD) : 0.36
menuju (CRD) : 0.26
lokasi (QTY) : 0.34
pada (PRD) : 0.33
internal untuk (PRD) : 0.31
pembukaan resmi berlangsung pada (PRD) : 0.41
di (CRD) : 0.75
dusun (DAT) : 0.24
gemp (CRD)

## Predict next words

### Subtask:
Prediksi 5 kata yang paling mungkin muncul setelah input satu kata tertentu menggunakan model N-gram.

In [38]:
from collections import Counter
import itertools
import nltk

# Flatten the list of lists of tokens into a single list
all_tokens = list(itertools.chain.from_iterable(df['processed_description']))

# Generate bigrams (pairs of consecutive words)
bigrams = list(nltk.bigrams(all_tokens))

# Count the frequency of each bigram
bigram_counts = Counter(bigrams)

# Generate trigrams (triplets of consecutive words)
trigrams = list(nltk.trigrams(all_tokens))

# Count the frequency of each trigram
trigram_counts = Counter(trigrams)

print("Bigram counts generated.")
print("Trigram counts generated.")

Bigram counts generated.
Trigram counts generated.


In [39]:
# Example usage: Predict the next 5 words after the word 'pengabdian' using bigram counts
input_word = 'pengabdian'
predicted_words_bigram = predict_next_words(input_word, bigram_counts)

print(f"Predicted next 5 words after '{input_word}' using bigram model:")
print(predicted_words_bigram)

# You can similarly use trigram_counts if you modify the predict_next_words function
# to handle trigrams or create a new function for trigram prediction.

Predicted next 5 words after 'pengabdian' using bigram model:
['masyarakat', 'hmti', 'desa', 'masyarakatpetani', 'inikondisi']


## Summary:

### Data Analysis Key Findings

* The dataset contains 30 articles with 'title' and 'description' fields.
* Text cleaning successfully removed HTML tags, URLs, numbers, and symbols, and converted text to lowercase.
* Text preprocessing involved tokenization, stopword removal (using Indonesian stopwords), and stemming (using Porter Stemmer), resulting in a processed list of tokens for each description.
* A TF-IDF matrix was generated, representing the articles as vectors. The matrix has a shape of (30, 1766), indicating 30 documents and 1766 unique words (features) after preprocessing.
* The top 10 TF-IDF words were successfully identified and displayed for each article, highlighting the most distinctive terms in each document.
* Attempts to perform POS tagging and NER using Stanza showed that while POS tagging and lemma extraction were successful, the NER model for Indonesian was not available, leading to an error.
* A bigram model was successfully built by flattening all processed tokens and counting the frequency of consecutive word pairs.
* The function `predict_next_words` was successfully implemented and demonstrated the ability to predict the most likely next words based on bigram frequencies for a given input word.

### Insights or Next Steps

* The limitation in performing NER highlights a potential area for further exploration: finding alternative Indonesian NLP libraries with robust NER capabilities or exploring the possibility of training a custom NER model.
* The bigram model provides a basic understanding of word sequences. For more advanced language modeling, building trigram or higher-order N-gram models, or utilizing deep learning models like LSTMs or Transformers, could be considered.
* The TF-IDF representation can be used for various downstream tasks such as document similarity analysis, clustering, or text classification.
* The insights gained from the top TF-IDF words can help understand the main themes and topics present in each article.